### DISCUSSION TITLE
setup_class does not work well with @pytest.fixture

### DISCUSSION THREAD

**Discussion by vipkum2** [pytest-dev/pytest#11330] [Q&A] [answered]
**What I want:** I need a session level setup for some data, this data need by the test scripts/tests, Also test scripts has some class level setup.
**Issue:** I am using below code to achieve the same but "setup_class" in test script, start executing with each test case. I am expecting it to run only once for the class. Without fixture It run once for class.

**Workaround:** I can use fixture directly in test case like: `def test_01(self,abc):`, it works. But I want that If I can only use value of "abc" in `setup_class`, so only one place I need to access the "abc".

```
#conftest.py file:

@pytest.fixture(scope="session")
def abc():
    print("\n\n ### ABC Session Level Setup ###")
    abc = 6
    #os.environ["a"] = str(data)
    yield abc
    print("\n\n ### ABC Session Level TearDown ###")
```

```
#test03.py
import pytest

class Test3:
    @pytest.fixture(autouse=True)
    def setup_class(self,abc):
        self.d = abc
        print("\n Inside Test3 setup_module",)

    def test_01(self,):
        print("test_01",self.d)

    def test_02(self,):
        print("test_02",self.d)

    def test_03(self,):
        print("test_03", self.d)```

```
#Execution logs:
vipin@nmjwr7pwg0 test % python3 -m pytest -s -vv test3.py -x
========================================================================= test session starts =========================================================================
platform darwin -- Python 3.10.6, pytest-7.2.0, pluggy-1.0.0 -- /opt/homebrew/opt/python@3.10/bin/python3.10
cachedir: .pytest_cache
rootdir: /Users/vipin/my_mac/pyLogger, configfile: pytest.ini
collected 3 items

test3.py::Test3::test_01

 ### ABC Session Level Setup ###

 Inside Test3 setup_module
test_01 6
PASSED
test3.py::Test3::test_02
 Inside Test3 setup_module
test_02 6
PASSED
test3.py::Test3::test_03
 Inside Test3 setup_module
test_03 6
PASSED

 ### ABC Session Level TearDown ###

========================================================================== 3 passed in 0.00s ==========================================================================
```

**Comment by The-Compiler** [comment]
Set the [fixture scope](https://docs.pytest.org/en/7.4.x/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) to `"class"` for that. Though, typically you would use the fixture directly in your test cases instead of storing things in the class instance.

**Reply by vipkum2** [reply]
@The-Compiler if I mark `scope="class"` then tests does not identify `setup_class` variables.
Got below errors:
```
self = <test3.Test3 object at 0x105a260e0>

    def test_01(self,):
>       print("test_01",self.d)
E       AttributeError: 'Test3' object has no attribute 'd'

test3.py:10: AttributeError
```

This is test script class:
 ```
@pytest.fixture(autouse=True, scope="class")
    def setup_class(self,abc):
        self.d = abc
```

**Reply by The-Compiler** [reply]
Works fine for me. Do you have a complete example that doesn't work?

**Reply by vipkum2** [reply]
The same example, I have just added scope=class in test03.py:
```
import pytest

class Test3:
    @pytest.fixture(autouse=True, scope="class")
    def setup_class(self,abc):
        self.d = abc
        print("\n Inside Test3 setup_module",)

    def test_01(self,):
        print("test_01",self.d)

    def test_02(self,):
        print("test_02",self.d)

    def test_03(self,):
        print("test_03", self.d)
```

**Reply by The-Compiler** [reply]
Can you please show a *complete* runnable example exhibiting the problem? If I use your example and add an `abc` fixture, it works as expected.

**Reply by The-Compiler** [reply]
Ah, my bad, this reproduces the issue:

```python
import pytest

@pytest.fixture(scope="class")
def abc():
    pass


class Test3:
    @pytest.fixture(autouse=True, scope="class")
    def a(self,abc):
        self.d = abc
        print("\n Inside Test3 setup_module",)

    def test_01(self,):
        print("test_01",self.d)

    def test_02(self,):
        print("test_02",self.d)

    def test_03(self,):
        print("test_03", self.d)
```

(but still - you make it much easier for people to help you if you post something they can run without any more modifications needed, see http://www.sscce.org/)

This indeed fails in the way you described, but that actually makes sense now that I think of it: Every test gets a new instance of your `Test3` class, so if you only add your attribute once, it follows that it doesn't exist for the other tests.

As said earlier, typically you'd use `d` (or whatever you'd call it) as a fixture instead of patching things onto the test instance.

**Reply by vipkum2** [reply]
I did not get the meaning of  line "as a fixture instead of patching things onto the test instance".
Do we have any workaround to achieve the same ?

**Reply by RonnyPfannschmidt** [reply, marked as answer]
for one, assign things to type(self).foo instead of self.foo

i believe we have a number of issues about needing to error on class scope fixtures in classes that take self

**Reply by vipkum2** [reply]
Thanks @RonnyPfannschmidt and @The-Compiler !
type(self) worked here.

**Reply by The-Compiler** [reply]
What I meant above, and what seems a way less hack-ish way to do things, is to use the fixture normally as an argument to your tests:

```python
class Test3:
    @pytest.fixture(scope="class")
    def abc(self):
        pass

    def test_01(self, abc):
        print("test_01", abc)

    def test_02(self, abc):
        print("test_02", abc)

    def test_03(self, abc):
        print("test_03", abc)
```

**Reply by vipkum2** [reply]
Yes, this is ideal way. In my case, we have already some scripts, so I am just avoiding to change each place. Thanks you !

**Reply by The-Compiler** [reply]
In that case, you can always have a class-scope fixture providing your value, but a function-scope fixture setting the attribute:

```python
import pytest

@pytest.fixture(scope="class")
def abc():
    pass


class Test3:
    @pytest.fixture(autouse=True)
    def a(self,abc):
        self.d = abc
        print("\n Inside Test3 setup_module",)

    def test_01(self,):
        print("test_01",self.d)

    def test_02(self,):
        print("test_02",self.d)

    def test_03(self,):
        print("test_03", self.d)
```

**Reply by cohml** [reply]
> typically you would use the fixture directly in your test cases instead of storing things in the class instance.

Do you mean you'd typically recommend this structure?

```python
@pytest.fixture(score="class")
def foo():
    return "foo"

class TestClass:
    def test(self, foo):
        ...  # use ``foo``
```

Can you explain why that is the recommended practice?

What is the drawback of defining a class-level fixture? It strikes me as quite tidy. E.g.,

```python
class TestClass:

    @pytest.fixture(scope="class")
    def foo(self):
        return "foo"

    def test(self, foo):
        ...  # use ``foo``
```

**Reply by The-Compiler** [reply]
That's not what I meant - if a fixture is only needed in that class, it should definitely be defined inside of it! What I meant is that you typically would use the fixture as an argument (like you did), and not do `self.foo = "foo"` and then use `self.foo` in the tests (like the OP did).

**Reply by cohml** [reply]
Gotcha. Re-reading your comment I see now what you meant, but thanks for clarifying. Glad to be on the right side of the law 😄
